smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/HttpProtocolTestGenerator.java

```diff
@@ -0,0 +1,484 @@
+
+            writer.write("const queryString = buildQueryString(r.query);");
+            explicitQueryValues.forEach(explicitQueryValue ->
+                    writer.write("expect(queryString).toContain($S);", explicitQueryValue));
+        }
+        writer.write("");
+    }
+
+    private void writeRequestHeaderAssertions(HttpRequestTestCase testCase) {
+        testCase.getRequireHeaders().forEach(requiredHeader ->
+                writer.write("expect(r.headers[$S]).toBeDefined();", requiredHeader));
+        writer.write("");
+
+        testCase.getForbidHeaders().forEach(forbidHeader ->
+                writer.write("expect(r.headers[$S]).toBeUndefined();", forbidHeader));
+        writer.write("");
+
+        testCase.getHeaders().forEach((header, value) -> {
+            writer.write("expect(r.headers[$S]).toBeDefined();", header);
+            writer.write("expect(r.headers[$S]).toBe($S);", header, value);
+        });
+        writer.write("");
+    }
+
+    private void writeRequestBodyAssertions(HttpRequestTestCase testCase) {
+        testCase.getBody().ifPresent(body -> {
+            // Fast fail if we don't have a body.
+            writer.write("expect(r.body).toBeDefined();");
+
+            // Otherwise load a media type specific comparator and do a comparison.
+            String mediaType = testCase.getBodyMediaType().orElse(null);
+
+            // Fast check if we expect an empty body, or we have an undescribed or plain text body.
+            if (body.isEmpty() || mediaType == null || mediaType.equals("text/plain")) {
+                writer.write("expect(r.body).toBe($S);", body);
+                return;
+            }
+            registerBodyComparatorStub(mediaType);
+
+            writer.write("const bodyString = `$L`;", body);
+            writer.write("const unequalParts: any = compareEquivalentBodies(bodyString, r.body.toString());");
+            writer.write("expect(unequalParts).toBeUndefined();");
+        });
+    }
+
+    private void registerBodyComparatorStub(String mediaType) {
+        // Load an additional stub to handle body comparisons for the
+        // set of bodyMediaType values we know of.
+        switch (mediaType) {
+            case "application/x-www-form-urlencoded":
+                additionalStubs.add("protocol-test-form-urlencoded-stub.ts");
+                break;
+            case "application/json":
+                additionalStubs.add("protocol-test-json-stub.ts");
+                break;
+            case "application/xml":
+                writer.addDependency(TypeScriptDependency.XML_PARSER);
+                writer.addImport("parse", "xmlParse", "fast-xml-parser");
+                additionalStubs.add("protocol-test-xml-stub.ts");
+                break;
+            default:
+                throw new CodegenException("Unable to compare bodies of media type `" + mediaType + "`");
```

Comment:
Should this throw or warn? Seems like adding a new comparison would be a breaking change to generators. I don't think we'd want to coordinate these updates across all implementations.

Reply:
Yeah -- trying an exact match + warning as a last resource seems good enough to me. At most, I think it could be configurable as to whether it fails or warns 🤷